src/editor/Editor.js

```diff
@@ -2385,6 +2401,133 @@ define(function (require, exports, module) {
+        this._codeMirror.refresh();
+
+        if (gutters.indexOf(LINE_NUMBER_GUTTER) < 0) {
+            $(rootElement).addClass("linenumber-disabled");
+        } else {
+            $(rootElement).removeClass("linenumber-disabled");
+        }
+    };
+
+    /**
+     * Sets the marker for the specified gutter on the specified line number
+     * @param   {string}   lineNumber The line number for the inserted gutter marker
+     * @param   {string}   gutterName The name of the gutter
+     * @param   {object}   marker     The dom element representing the marker to the inserted in the gutter
+     */
+    Editor.prototype.setGutterMarker = function (lineNumber, gutterName, marker) {
+        var gutterNameRegistered = registeredGutters.some(function (gutter) {
+            return gutter.name === gutterName;
+        });
+
+        if (!gutterNameRegistered) {
+            console.warn("Gutter name must be registered before calling editor.setGutterMarker");
+            return;
+        }
+
+        this._codeMirror.setGutterMarker(lineNumber, gutterName, marker);
+    };
+
+    /**
+     * Returns the list of gutters current registered on all editors.
+     * @return {!Array.<{name: string, priority: number}>}
+     */
+    Editor.getRegisteredGutters = function () {
+        return registeredGutters;
+    };
+
+    /**
+     * Registers the gutter with the specified name at the given priority.
+     * @param {string} name    The name of the gutter.
+     * @param {number} priority  A number denoting the priority of the gutter. Priorities higher than LINE_NUMBER_GUTTER_PRIORITY appear after the line numbers. Priority less than LINE_NUMBER_GUTTER_PRIORITY appear before.
+     * @param {?Array<string>} languageIds A list of language ids that this gutter is valid for. If no language ids are passed, then the gutter is valid in all languages.
+     */
+    Editor.registerGutter = function (name, priority, languageIds) {
+        if (isNaN(priority)) {
+            console.warn("A non-numeric priority value was passed to registerGutter. The value will default to 0.");
+            priority = 0;
+        }
+
+        if (!name || typeof name !== "string") {
+            throw new Error("The name of the registered gutter must be a string.");
+        }
+
+        var gutter = {name: name, priority: priority, languages: languageIds},
+            gutterExists = registeredGutters.some(function (gutter) {
+                return gutter.name === name;
+            });
+
+        if (!gutterExists) {
+            registeredGutters.push(gutter);
+        }
+
+        Editor.forEveryEditor(function (editor) {
+            editor._renderGutters();
+        });
+    };
+
+    /**
+     * Unregisters the gutter with the specified name and removes it from the UI.
+     * @param {string} name The name of the gutter to be un registered.
+     */
+    Editor.unregisterGutter = function (name) {
+        var i, gutter;
```

Comment:
Maybe would prefer `Array.prototype.filter` for simplicity over couple of possible cycles saved 🤔 😄